### PR TITLE
feat: directional nlist

### PR DESCRIPTION
**Pull request by wanghan-iapcm**
this pr implement the directional nlist, with which the central and neighboring atoms are treated as different atoms. 

## Summary by CodeRabbit

- **New Features**
	- Introduced a new function for building a directional neighbor list, enhancing detailed neighbor analysis.
- **Bug Fixes**
	- Improved handling of neighbor lists, ensuring proper management of virtual atoms and accurate sizing.
- **Tests**
	- Added new tests for the directional neighbor list functionality and modified existing tests to cover additional scenarios.

### CODE REVIEW

**Comment by wanghan-iapcm** [comment]
crashed env on my local computer, the UTs will be fixed later on

File: deepmd/pt/utils/nlist.py [hunk 1]
```diff
@@ -126,8 +125,26 @@ def build_neighbor_list(
...
+
+def _trim_mask_distinguish_nlist(
+    is_vir_cntl,
+    atype_neig,
+    rr,
+    nlist,
+    rcut,
+    sel,
+    distinguish_types,
+):
```
Comment (iProzd) [inline deepmd/pt/utils/nlist.py:142]:
```suggestion
def _trim_mask_distinguish_nlist(
    is_vir_cntl: torch.Tensor,
    atype_neig: torch.Tensor,
    rr: torch.Tensor,
    nlist: torch.Tensor,
    rcut: float,
    sel: List[int],
    distinguish_types: bool,
):
```

Comment (iProzd) [reply deepmd/pt/utils/nlist.py:142]: This may help to pass jit UT

**Review by iProzd — CHANGES_REQUESTED** [review]

File: deepmd/pt/utils/nlist.py (hunk 1 above)
Comment (wanghan-iapcm) [reply deepmd/pt/utils/nlist.py]: thanks!

File: deepmd/pt/utils/nlist.py [hunk 2]
```diff
@@ -126,8 +125,26 @@ def build_neighbor_list(
...
+    rcut: float,
+    sel: List[int],
+    distinguish_types: bool,
+):
+    """Trim the size of nlist, mask if any central atom is virtual, distinguish types if necessary."""
+    nsel = sum(sel)
     # nloc x nsel
-    nnei = rr.shape[2]
+    batch_size, nloc, nnei = rr.shape
+    assert batch_size, nloc == is_vir_cntl.shape
```
Comment (njzjz) [inline deepmd/pt/utils/nlist.py:147]: I believe that `==` has higher precedence than `,`

Comment (njzjz) [reply deepmd/pt/utils/nlist.py:147]:
```py
>>> 1,2==(1,2)
(1, False)
```

**Comment by njzjz** [comment]
What is the purpose of introducing a different neighbor list?

**Comment by wanghan-iapcm** [comment]
> What is the purpose of introducing a different neighbor list?

For the development of charge density models.

**Review by iProzd — APPROVED** [review]

**Review by njzjz — APPROVED** [review]

**Pull request merged**